src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs

```diff
@@ -623,6 +626,11 @@ private bool TakeStartLine(SocketInput input)
                     return false;
                 }
 
+                char[] requestUriChars;
+                var rawPathLength = beginPath.GetCharArray(endPath, out requestUriChars);
+                var decodedLength = UrlPathDecoder.DecodeInPlace(requestUriChars, rawPathLength);
```

Comment:
All right. I'm working out a prototype on this idea now. I'll send a separate PR for that. But I'd like to keep this floating as a fall back.